### DISCUSSION TITLE
Zustand always rerenders the whole page when used in Apollo Query affecting infinite loop

### DISCUSSION THREAD

**Discussion by wisnupramoedya** [pmndrs/zustand#1940] [Q&A]
Hi everyone. I'm having trouble using Zustand on Apollo Query. When I log in, when Zustand tries to change its state, the whole page suddenly renders in full. This causes Apollo Query to try to call the API connection again, and Zustand also changes state again. I've tried the solution on [Create Problem](https://github.com/pmndrs/zustand/discussions/372) as well as [using useEffect](https://github.com/apollographql/apollo-client/issues/9454). Is there something wrong with my understanding of set state?

Here is the code for Zustand. I use immer and create selector hooks.
```
import { createSelectorHooks } from 'auto-zustand-selectors-hook';
import produce from 'immer';
import create from 'zustand';

/* This is just a type variable */
import { Team } from '@/gql/graphql';

type TeamStoreType = {
  team: Team | null;
  setTeam: (team: Team) => void;
};

const useTeamStoreBase = create<TeamStoreType>((set, get) => ({
  team: null,
  setTeam: (team) => {
    if (get().team !== team) {
      console.log({ message: 'SET TEAM' });

      set(
        produce<TeamStoreType>((state) => {
          state.team = team;
        })
      );
    }
  },
}));

const useTeamStore = createSelectorHooks(useTeamStoreBase);

export default useTeamStore;
```

Here is the code for my component.
```
import { useLazyQuery } from '@apollo/client';
import { useRouter } from 'next/router';
import { useEffect } from 'react';

import useTeamStore from '@/stores/useTeamStore';

import { graphql } from '@/gql';
import { Team } from '@/gql/graphql';

const QUERY = graphql(/* GraphQL */ `
  query GetTeamStoreDetail($teamId: Int) {
    team(id: $teamId) {
      id
      title
    }
  }
`);

/**
* Component works by querying every time the slug is changing. The results will be passed to useTeamStore.
*/
const Component = () => {
  const router = useRouter();
  const { slug } = router.query;

  const setTeam = useTeamStore.useSetTeam();
  const team = useTeamStore.useTeam();

  const { data, loading, error } = useQuery(QUERY, {
    variables: {
      teamId: parseInt(slug as string),
    },
    onCompleted: () => {
      return;
    },
  });

  useEffect(() => {
    if (data) {
      setTeam(data.team as Team);
    }
  }, [data]);

 return (
    <div className='grid grid-cols-1 gap-y-5 gap-x-9'>
      ...
    </div>
  );
};
  
```

**Comment by nikhcevov** [comment]
Hmm, did you tried to set state in `onCompleted`?

```
 const { data, loading, error } = useQuery(QUERY, {
    variables: {
      teamId: parseInt(slug as string),
    },
    onCompleted: (data) => {
      setTeam(data.team as Team);
    },
  });
```

**Reply by dbritto-dev** [reply]
@wisnupramoedya the right way to integrate `Apollo GraphQL` with `zustand` is by using the `client` instance directly. For instance:

```ts
import { createSelectorHooks } from 'auto-zustand-selectors-hook';
import produce from 'immer';
import create from 'zustand';

import { graphql } from '@/gql';
import { Team } from '@/gql/graphql';

/* This is just a type variable */
import { Team } from '@/gql/graphql';

type TeamStoreType = {
  team: Team | null;
  setTeam: (team: Team) => void;
  fetchTeam: (teamId: number) => Promise<void>;
};

const useTeamStoreBase = create<TeamStoreType>((set, get) => ({
  team: null,
  setTeam: (team) => {
    set(
      produce<TeamStoreType>((state) => {
        state.team = team;
      })
    );
  },
  fetchTeam: async (teamId: number) => {
    const result = await client.query({
      query: graphql(/* GraphQL */ `
        query GetTeamStoreDetail($teamId: Int) {
          team(id: $teamId) {
            id
            title
          }
        }
      `),
      variables: { teamId }
    })
    
    get().setTeam(result.data)
  }
}));

const useTeamStore = createSelectorHooks(useTeamStoreBase);

export default useTeamStore;
```

> **Note:** this is not the final/real solution but just an example. The final solution would require a bunch of code in order to keep everything in sync. We recommend to keep your data on your data layer, react query, or apollo.

**Comment by jerelmiller** [comment]
Hey all 👋 

Apollo Client maintainer here. I came across this thread and wanted to chime in with a couple things to be aware of, especially in regards to the [proposed solution](https://github.com/pmndrs/zustand/discussions/1940#discussioncomment-7732181). 

tl;dr; We recommend keeping data loaded via Apollo Client separate from your other application data stored in Zustand and avoid trying to sync the two stores. Let Apollo Client be your single source of truth for GraphQL data and let it sync data to your components.

---

Apollo Client's cache is a data store, which means that storing that data in Zustand requires you to keep both stores in sync with each other. You risk reading outdated data in your application otherwise. 

This is especially problematic if you have two queries that return overlapping data. Apollo Client's `InMemoryCache` is a [normalized data store](https://www.apollographql.com/docs/react/caching/overview#data-normalization) which handles overlapping data from multiple queries quite well because it dedupes the overlapping data between different queries into a single JavaScript object. Changes to those records can be done in a single location and be kept up-to-date by Apollo Client's cache and propagate those changes to your components. That is the job of the hooks APIs (`useQuery`, `useSuspenseQuery`, etc.) and why you don't see `client.query(...)` used more often in components directly.

While the `client.query(...)` call in the proposed solution works to get your initial query data, you have no way of receiving updates from the cache, which means changes to that data via mutations, other queries, or manual cache writes, will be missed by your components. You'll find that in order to keep these two in sync, you end up having to write a bunch of code your self or design your Zustand store(s) to be able to handle edge cases that Apollo Client already handles for you.

Take this example:

```gql
# One query loads my list of todos
query MyTodos {
  me {
    todos {
      id
      title
      completed
    }
  }
}

# Another loads a single todo
query GetTodo($id: ID!) {
  todo(id: $id) {
    id
    title
    completed
  }
}

# A mutation that lets us complete the todo
mutation CompleteTodo($id: ID!) {
  completeTodo(id: $id) {
    id
    completed
  }
}
```

And data returned from each query as such:

```js
// MyTodos
{
  "me": {
    __typename: 'User',
    "todos": [
      { __typename: 'Todo', id: 1, title: 'Clean room', completed: false },
      { __typename: 'Todo', id: 2, title: 'Walk dog', completed: true }
    ]
  }
}

// GetTodo
{
  todo: {
    __typename: 'Todo',
    id: 1,
    title: 'Clean room',
    completed: false
  }
}
```

You'll note how the two different queries return the same `Todo` record (the `id` of `1`). Apollo Client normalizes this data so it will be stored in the cache like such:

```js
{
  ROOT_QUERY: {
    me: {
      __typename: 'User',
      todos: [
        { __ref: 'Todo:1' },
        { __ref: 'Todo:2' }
      ] 
    }
    'todo({"id":1})': {
      __ref: 'Todo:1'
    }
  },
  'Todo:1': {
    __typename: 'Todo',
    id: 1,
    title: 'Clean room',
    completed: false 
  },
  'Todo:2': {
    __typename: 'Todo',
    id: 2,
    title: 'Walk dog',
    completed: true
  }
}
```

Now let's run the `CompleteTodo` mutation to complete the `Todo` with `id` of `1`:

```js
const [completeTodo] = useMutation(COMPLETE_TODO)

// in a handler somewhere:
completeTodo({ variables: { id: 1 } })
```

Apollo Client will see that the data returned from your mutation affects `Todo:1` and will update it:

```js
{
  'Todo:1': {
    __typename: 'Todo',
    id: 1,
    title: 'Clean room',
    completed: true // <-- This is now completed
  }
}
```

Because the `Todo` is normalized, Apollo Client will notify any queries that contain this data that this record has been updated and will rerender your components for you, meaning the `MyTodos` and `GetTodo` queries will contain the updated `completed` flag for `Todo:1`.

If you're syncing this data to Zustand, this process becomes more difficult because you'll need to write a bunch of boilerplate code to sync these two. The difficulty of this may depend on the shape of your Zustand data store and how you transform the data from your GraphQL query into Zustand. If you're not doing some kind of normalization yourself in Zustand (i.e. extracting the common `Todo` records out to a single record), you risk missing an update to a copy of `Todo:1` somewhere in your store which means that copy may now contain outdated data resulting in UI bugs. I think you'll find that writing this syncing logic yourself can be pretty cumbersome and prone to errors if you're not careful.

As such, we don't recommend storing your GraphQL data in Zustand and instead keep that data in Apollo Client as your single source of truth, otherwise you either risk reading outdated data somewhere in your app.

PS - if you're wanting to read a piece of data from the Apollo Client cache without having to run it through an entire query, check out [`useFragment`](https://www.apollographql.com/docs/react/data/fragments#usefragment) which provides a lightweight binding to the cache for a selection of data on a normalized record. This hook will keep your components in sync with the cache.

Hope this helps!

**Reply by dbritto-dev** [reply]
@jerelmiller I'm down with you but  my proposed solution is not the solution to the problem per se, is just an example. The real solution as you mentioned require to add a bunch of code to keep everything in sync and we do the same recommendation when people ask to keep the data on zustand even when they use solutions like `react-query` and `apollo`

**Reply by dbritto-dev** [reply]
@jerelmiller I updated my comment to note that

**Reply by jerelmiller** [reply]
@dbritto-dev no problem! Thanks for adding that amendment. I just wanted to make sure to communicate the trade-offs in case it wasn't obvious why it is/isn't a good idea to sync the two since I didn't see it elsewhere (granted I didn't look very hard).

**Reply by dbritto-dev** [reply]
@jerelmiller idk why people keep asking about moving data from data layers to zustand. I guess we need to talk more about server states, client states and component states. 

BTW, I'm working on revamping docs to tackle api ref and recipes in general with troubleshooting sections

**Reply by jerelmiller** [reply]
Haha all good. I'd say Apollo's docs don't do a super great job of stating that its cache is itself a data store, so I think the "cache" part of the name tends to imply its just there to speed up future requests. I think its probably half our fault why devs might try to integrate the two if they don't know better.
